hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryClientStateRegistry.java

```diff
@@ -199,36 +207,56 @@ public void closeOnError(QueryId queryId) {
     private void close0(QueryClientState clientCursor) {
         clientCursor.getSqlResult().close();
```

Comment:
`getSqlResult` might return `null`. I couldn't find a specific scenario, but I'm not 100% sure it's guaranteed to be non-null. Or an edge case can be added in the future. It won't do any harm here.

Reply:
`getSqlResult` returns `null` when the query was canceled before the execution is started. This also explains why we need the query ID here https://github.com/hazelcast/hazelcast/pull/18128#discussion_r568523961

Reply:
I agree it probably can't happen in the current code. But `close` methods should be written more defensively. They are typically idempotent and they are assumed that they can be called in any state - they are often called in `finally` blocks. For example Mysql JDBC driver has an issue that `disconnect` fails if an XA transaction is open - you're doing a txn, you fail, you want to disconnect but you can't...

If we don't change this, maybe it would be better to name this method `closeIfNotCancelledBeforeExecute`

Reply:
Looks like I mixed this and the previous comments. Yes, it is better to have a null-check here, fixed.